kernel/h1/src/hil/flash/h1_hw.rs

```diff
@@ -265,27 +266,56 @@ impl super::hardware::Hardware for H1bHw {
         self.error_code.get() as u16
     }
 
-    fn set_transaction(&self, offset: usize, size: usize) {
+    fn set_transaction(&self, offset: usize, size: usize) -> ReturnCode {
         use self::TransactionParameters::{Offset,Size};
-        // The offset is relative to the beginning of the flash module. There
-        // are 128 pages per flash module.
-        // TODO(jrvanwhy): Assumes the read is from the second flash bank.
-        if offset > H1_FLASH_SIZE {
-           return; // TODO(pal): Fails silently!
+        // The offset is relative to the beginning of the flash module.
+        if offset > H1_FLASH_SIZE / BYTES_PER_WORD {
+           return ReturnCode::EINVAL
+        }
+
+        if size > H1_FLASH_BANK_SIZE / BYTES_PER_WORD {
+            return ReturnCode::EINVAL
+        }
+
+        // Compute the offset relative to the flash bank.
+        let relative_offset: usize;
+        if offset < H1_FLASH_BANK_SIZE / BYTES_PER_WORD {
+            relative_offset = offset;
+        } else {
+            relative_offset = offset - H1_FLASH_BANK_SIZE / BYTES_PER_WORD;
+        }
+
+        if relative_offset + size > H1_FLASH_BANK_SIZE / BYTES_PER_WORD {
+            return ReturnCode::EINVAL
```

Comment:
This is a lot of complexity for untestable code. Can some of this logic be moved out of `H1bHw`?

Reply:
Good point. I moved that logic out into the `Driver`.

Reply:
If you like I can squash the two commits and force push if that makes it easier to read.